src/ast.rs

```diff
@@ -592,6 +592,7 @@ pub enum BindingKind {
     Let,
     Assert,
     Try,
+    Clause,
```

Comment:
This doesn't seem right to me. It's now possible to construct a binding with a `clause` kind, which would render invalid Gleam in the formatter. I think as much as possible we could not have data structures that permit invalid states.

Reply:
That seems reasonable. I'm trying to use it for two purposes I guess, and it would probably be better to keep them separate. Are you happy with the general idea of putting this logic in the `PatternTyper` with the extra binding_kind field? If so I could make a new enum for this purpose which would leave BindingKind unchanged.

```rust
// Not sure about the name yet!
enum PatternUsage {
    Binding(BindingKind)
    Clause
}
```

Reply:
https://github.com/gleam-lang/gleam/blob/d7b2fb66ecfb8c4fa7cbf32b1143f81bc9320568/src/typ.rs#L1630-L1637

This code here makes me think that it's not the right place. The pattern typer is being constructed to handle a single clause rather than the entire case expression, so if later this code was upgrades to do full case expressions we would have to remove it from there and place it at some higher level.

I'm not sure where it should go, but I think it should go somewhere that _could_ be the home of full exhaustiveness checking in future, at least or `let` and `try`.

Reply:
The issue is that moving this code to outside the `PatternTyper` (back to where I had it originally in the `infer_let()` for example) is that it won't handle nested constructors properly. I think it makes sense to be checking the exhaustiveness as you infer the Pattern, otherwise you'll just have to go back through it all again afterwards. Putting it back into `infer_let()` would also not help at all with the hypothetical future full exhaustiveness checking, as the let version is really just a single clause version of the generalised case one.

Possibly a solution would be to build up a data structure that represents the coverage state as we recursively infer the pattern elements, and then return that alongside the final `TypedPattern` ast. A separate bit of code could then be called from `infer_let()` and `infer_case()` with that data to actually decide if the coverage state is exhaustive or not. This might even start looking like a bit like the Drop Your Guards algorithm from that paper @QuinnWilton posted in the original exhaustiveness issue #34. Or maybe I just think that because I spent a chunk of this morning trying (and mostly failing) to get my head round said paper.

Not sure that helps us much in the short term though! I'll see if anything clever suddenly occurs to me.

Reply:
Let me know when you have an idea of what you would like to do (including continuing with it in the pattern typer) and I'll do a full review. Thanks!